dronegen/os_repos.go

```diff
@@ -0,0 +1,460 @@
+	optpb := &OsPackageToolPipelineBuilder{
+		clameName:          claimName,
+		packageType:        packageType,
+		packageManagerName: packageManagerName,
+		bucketSecrets:      bucketSecrets,
+		extraArgs:          []string{},
+		setupCommands:      []string{},
+		requiredPackages:   []string{},
+		volumeName:         fmt.Sprintf("%s-persistence", packageManagerName),
+		pipelineNameSuffix: fmt.Sprintf("%s-new-repos", packageManagerName),
+		artifactPath:       "/go/artifacts",
+		pvcMountPoint:      "/mnt",
+	}
+
+	optpb.environmentVars = map[string]value{
+		"REPO_S3_BUCKET": {
+			fromSecret: optpb.bucketSecrets.bucketName,
+		},
+		"AWS_ACCESS_KEY_ID": {
+			fromSecret: optpb.bucketSecrets.accessKeyID,
+		},
+		"AWS_SECRET_ACCESS_KEY": {
+			fromSecret: optpb.bucketSecrets.secretAccessKey,
+		},
+		"AWS_REGION": {
+			raw: "us-west-2",
+		},
+		"BUCKET_CACHE_PATH": {
+			// If we need to cache the bucket on the PVC for some reason in the future
+			// uncomment this line
+			// raw: path.Join(pvcMountPoint, "bucket-cache"),
+			raw: "/tmp/bucket",
+		},
+		"ARTIFACT_PATH": {
+			raw: optpb.artifactPath,
+		},
+		"GNUPGHOME": {
+			raw: "/tmpfs/gnupg",
+		},
+		"GPG_RPM_SIGNING_ARCHIVE": {
+			fromSecret: "GPG_RPM_SIGNING_ARCHIVE",
+		},
+		"DEBIAN_FRONTEND": {
+			raw: "noninteractive",
+		},
+	}
+
+	return optpb
+}
+
+func (optpb *OsPackageToolPipelineBuilder) buildPromoteOsPackagePipeline() pipeline {
+	pipelineName := fmt.Sprintf("publish-%s", optpb.pipelineNameSuffix)
+	checkoutPath := "/go/src/github.com/gravitational/teleport"
+	commitName := "${DRONE_TAG}"
+	checkoutStepName := "Check out code"
+
+	p := optpb.buildBaseOsPackagePipeline(pipelineName, checkoutStepName, checkoutPath, commitName)
+	p.Trigger = triggerPromote
+	p.Trigger.Repo.Include = []string{"gravitational/teleport"}
+
+	setupSteps := []step{
+		{
+			Name:  "Verify build is tagged",
+			Image: "alpine:latest",
+			Commands: []string{
+				"[ -n ${DRONE_TAG} ] || (echo 'DRONE_TAG is not set. Is the commit tagged?' && exit 1)",
+			},
+		},
+	}
+	setupSteps = append(setupSteps, p.Steps...)
+	setupSteps = append(setupSteps,
+		step{
+			Name:  "Check if tag is prerelease",
+			Image: "golang:1.17-alpine",
+			Commands: []string{
+				fmt.Sprintf("cd %q", path.Join(checkoutPath, "build.assets", "tooling")),
+				"go run ./cmd/check -tag ${DRONE_TAG} -check prerelease || (echo '---> This is a prerelease, not publishing ${DRONE_TAG} packages to APT repos' && exit 78)",
+			},
+		},
+	)
+
+	setupStepNames := make([]string, 0, len(setupSteps))
+	for _, setupStep := range setupSteps {
+		setupStepNames = append(setupStepNames, setupStep.Name)
+	}
+
+	versionSteps := optpb.getDroneTagVersionSteps(checkoutPath)
+	for i := range versionSteps {
+		versionStep := &versionSteps[i]
+		if versionStep.DependsOn == nil {
+			versionStep.DependsOn = setupStepNames
+			continue
+		}
+
+		versionStep.DependsOn = append(versionStep.DependsOn, setupStepNames...)
```

Comment:
What's the use of `DependsOn` if things are all listed sequentially anyways? Is it for future plans to parallelize things?

Reply:
Yep that's exactly it
I decided the trade off in this specific places wasn't worth the increased code complexity (YUM publishing is fast). However there are a few other places that this may be able to greatly decrease run time